[NDSL] pyMoist GFDL Single Moment & Python <--> Fortran bridge infrastructure

CharlesKrop opened the pull request:
Primary work: Pulled together the two main parts of the GFDL_1M module (phase change loop and driver) into a single class (`GFDL1M`), adding necessary glue code and supporting setup/finalization code. Tied this new GFDL1M class into the fortran with the creation of a new MAPLPy module that enables Fortran <--> Python memory management/exchange.

Other changes:

- removed references to and uses of the old python bridge
- refactored qsat functions to a more gt4py friendly structure and eliminate extra calculations - these changes may cause issues elsewhere in pyMoist
- minor changes to other pyMoist functions (revised field names, added documentation)
- removed old pymoist files/functions which were no longer being used

Code has been tested at C24 and runs for 24 hours without issue. Each of the components (supporting code, driver, and phase change loop) were tested with individual translate tests. Due to the complexity of the MAPL integration in the Fortran code, construction of a unified translate test was skipped. Translate tests revealed minor numerical differences originating from a documented Fortran v Python computation difference (see https://github.com/NOAA-GFDL/NDSL/issues/173 for more information).

TODO: radar diagnostic output still has to be ported. These fields are not re-used within the rest of the model, and their absence therefore has no effect on the validity of this code.

CharlesKrop committed:
progress on reworking phase change loop

CharlesKrop committed:
more progress on fixing the phase change loop

CharlesKrop committed:
Initial work fixing qsat tables

CharlesKrop committed:
more qsat work

CharlesKrop committed:
more qsat

CharlesKrop committed:
Working GFDL_1M phase change loop.

Working (new) pyMoist saturation tables module

CharlesKrop committed:
Updated docstrings, new GFDL_1M main call, minor reformats to all translate tests and phase change loops. linting

FlorianDeconinck committed:
Merge branch 'dsl/experimental/MAPLpyish' into dsl/GFDL_1M

FlorianDeconinck committed:
MAPLPyish: Fix compile of C bridge

CharlesKrop committed:
More work on GFDL unified call

CharlesKrop committed:
more GFDL_1M single call updates

FlorianDeconinck committed:
MAPLpy: Clean up API calls, introduce proper types, add spec for int32 get_resource

FlorianDeconinck committed:
GitIgnore the usual

FlorianDeconinck committed:
Clean up type hints

FlorianDeconinck committed:
Introducing user facing MAPLMemoryFactory

FlorianDeconinck committed:
Unused parameter

FlorianDeconinck committed:
Expose fortran `associated`

FlorianDeconinck committed:
MAPLPyish: fix GetResource API

FlorianDeconinck committed:
MAPLpyish: Fix GetPointer call through API

FlorianDeconinck committed:
Interface: pass down IMPORT, EXPORT, INTERNAL, MAPL_COMP
Base hook for GFDL with MAPL and example

FlorianDeconinck committed:
Hooks in Fortran

CharlesKrop committed:
Single call constructed and verified *in pieces*. All parts (setup, phase change, driver, finalize) verify (or have documented sources of errors) on their own. A unified translate test cannot be easily constructed. If needed, this will be reapproached at a later time

CharlesKrop committed:
Merge branch 'dsl/GFDL_1M_single_call' of github.com:CharlesKrop/GEOSgcm_GridComp into dsl/GFDL_1M_single_call

CharlesKrop committed:
Removing file and folder that should not have been committed

CharlesKrop committed:
Work towards integration

CharlesKrop committed:
work towards integration

FlorianDeconinck committed:
MAPL: Add 2D GetPointer, error for anything not 2D or 3D

CharlesKrop committed:
Successful initalization of GFDL_1M single call and linking of memory

CharlesKrop committed:
Functional single call infrastrucure

CharlesKrop committed:
removing extra files

CharlesKrop committed:
First bit of pre-commit changes

CharlesKrop committed:
Interface updates

CharlesKrop committed:
Added validation of dissapative_ke_heating function. precommit linting

CharlesKrop committed:
more precommit linting

CharlesKrop committed:
minor corrections

CharlesKrop committed:
fixing typo

CharlesKrop committed:
removing unnecessary print statements

FlorianDeconinck commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/interface/flags.py:
```diff
@@ -25,89 +25,88 @@ class MoistFlags:
 @dataclass
 class GFDL1MFlags:
     # GFDL_1M driver configuration. Initial values are not true defaults.
```
Probably a TODO here to make initial values true defaults

FlorianDeconinck reviewed (changes requested):
First batch

FlorianDeconinck committed:
Lint missing types

FlorianDeconinck committed:
Swap `condensate` temps for direct return

CharlesKrop committed:
Changes for the PR

CharlesKrop committed:
Merge branch 'dsl/GFDL_1M_single_call' of github.com:CharlesKrop/GEOSgcm_GridComp into dsl/GFDL_1M_single_call

CharlesKrop commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/GFDL_1M/PhaseChange/masks.py:
```diff
@@ -0,0 +1,18 @@
+
+@dataclass
+class Masks:
+    k_mask: Quantity
```
was unused. just didn't delete

CharlesKrop commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/GFDL_1M/PhaseChange/outputs.py:
```diff
@@ -0,0 +1,23 @@
+
+    @classmethod
+    def make(cls, quantity_factory: QuantityFactory):
+        # -----------------------------------------------------------------------
```
oh hah. back then I was copying the fortran style

CharlesKrop commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/GFDL_1M/config.py:
```diff
@@ -1,10 +1,10 @@
 
 @dataclass
-class MicrophysicsConfiguration:
+class GFDL1MConfig:
```
I don't absolute confidence in my knowledge of them either. There is no documentation in the fortran for the vast majority of them, so I'm going off of names and context clues. Those are fine for some, but not nearly sufficient for others, and I'm wary of solidifying docstrings without proper definitions (I'll try to get them from Fortran people)

CharlesKrop commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/GFDL_1M/driver/warm_rain/main.py:
```diff
@@ -175,18 +176,23 @@ def __call__(
-            precip_fall,
-        )
+        if self.GFDL_1M_config.USE_PPM is False:
+            # NOTE: somehow errors pop up in rain1 and m1_sol within implicit fall, despite all of the
```
Potentially, but not necessarily. I struggled mightily to figure this out but I'm rather stumped. It's a rather long explanation so I'm not going to put it here, but at some point we can chat about it if you want to try to solve it.

CharlesKrop commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/GFDL_1M/outputs.py:
```diff
@@ -0,0 +1,194 @@
+
+
+@dataclass
+class Outputs:
```
added a bit but not sure it is sufficient. It's going to be supremely difficult to get units for all of these like I did in `state.py`, and I tried to make the names as self-explanatory as possible, so there's not really anything to add to them.

CharlesKrop committed:
more PR changes

CharlesKrop committed:
Fixing build errors

CharlesKrop committed:
Correct warning for missing radar output

CharlesKrop committed:
Merge branch 'dsl/GFDL_1M_single_call' of github.com:CharlesKrop/GEOSgcm_GridComp into dsl/GFDL_1M_single_call

FlorianDeconinck commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/GEOS_GFDL_1M_InterfaceMod.F90:
```diff
@@ -408,9 +411,36 @@ subroutine GFDL_1M_Run (GC, IMPORT, EXPORT, CLOCK, RC)
     call ESMF_TimeIntervalGet(TINT,   S_R8=DT_R8,RC=STATUS); VERIFY_(STATUS)
     DT_MOIST = DT_R8
 
+#ifdef PYMOIST_INTEGRATION
```
Let's wait for this.

FlorianDeconinck commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/savepoint/GFDL_1M/PhaseChange/translate_phase_change.py:
```diff
@@ -0,0 +1,183 @@
+from pyMoist.GFDL_1M.PhaseChange.phase_change import PhaseChange
+
+
+class Translatephase_change(TranslateFortranData2Py):
```
Renamed to `TranslatePhaseChange` (and rename serialbox savepoint)

FlorianDeconinck committed:
Add print options to TimerCudaProfiler

FlorianDeconinck committed:
Fix MAPLPy brisge to properly use `value` for pointer passing

FlorianDeconinck committed:
Pass correct INTERNAL down
Add trampoline function for malfunctioning INTERNAL state at GFDL interface level

FlorianDeconinck committed:
Adapt wrapper to new INTERNAL state flow. Lint.

FlorianDeconinck committed:
Merge remote-tracking branch 'CharlesKrop/dsl/GFDL_1M_single_call' into dsl/GFDL_1M_single_call

FlorianDeconinck committed:
Fix bad API passthrough

FlorianDeconinck committed:
Move GFDL_1M to use parameters for runtime
GFDL_1M.__init__ gives public API
Pass optional fields as optional
Type hints & docs

FlorianDeconinck committed:
Add evapc/sublc as outputs to GFDL_1M
PhaseChange no longers as dataclass Outputs, but simply pass them in

FlorianDeconinck committed:
Adapt wrapper to revamp API
Clean some lint issues

FlorianDeconinck committed:
Orchestrate + Lint

FlorianDeconinck committed:
Remove unused code

CharlesKrop committed:
First attempt at making gt4py complient saturation tables. Still disabled for reasons noted in main.py

CharlesKrop committed:
Minor bug fixes and cleanup to prepare for PR

CharlesKrop committed:
Fix one issue preventing saturation tables from being constructed using gt4py stencils

CharlesKrop committed:
Fix one issue preventing saturation tables from being constructed using gt4py stencils

CharlesKrop committed:
Merge branch 'dsl/GFDL_1M_single_call' of github.com:CharlesKrop/GEOSgcm_GridComp into dsl/GFDL_1M_single_call

CharlesKrop committed:
tiny interface fix

CharlesKrop committed:
minor fix removing left over debug lines

CharlesKrop committed:
Compatability with NDSL milestone2 branch

CharlesKrop committed:
Removing DS_Store

CharlesKrop committed:
Added a switch to control how the saturation tables are created

FlorianDeconinck committed:
Break vertical interpolation

CharlesKrop committed:
Removed externals from saturation tables

CharlesKrop committed:
New vertical interpolation is now zero diff from the old version

CharlesKrop committed:
Minor change to vertical_interpolation_interface

FlorianDeconinck committed:
Driver tables: replace conditional on K with interval

FlorianDeconinck committed:
Fix first melted level mask + remove redundant mask reset at top level

CharlesKrop committed:
warm_rain bugfix: zero m1_rain correctly in step1 and pass it to implicit_fall

FlorianDeconinck committed:
Warm rain fix: implicit fall bad arguments, need to pass dud m1 temporaries unlike in terminal fall that uses that m1

CharlesKrop committed:
New driver setup & finish translate tests, minor fix in setup

CharlesKrop committed:
Fixed the default value for CCW_EVAP_EFF and CCI_EVAP_EFF

CharlesKrop committed:
Correctly tied the triggers on evaporate and sublimate to the do_evap and do_subl flags

FlorianDeconinck committed:
Shift `PFL_LS`, `PFI_LS`, `PFL_AN`, `PFI_AN ` to be true 72 level fields (see finalize.py)

FlorianDeconinck committed:
GFDL1M temporaries & mask are badly reset (one of them) - we hard reset all

FlorianDeconinck committed:
Interface: log cast failure, un-associated field transfer

FlorianDeconinck committed:
Merge remote-tracking branch 'CharlesKrop/dsl/GFDL_1M_single_call' into dsl/GFDL_1M_single_call

FlorianDeconinck committed:
Interface: introduce status check in bridge code

FlorianDeconinck committed:
Lint (basic pass)

FlorianDeconinck committed:
Hydrostatic PDF: lint + type hint a single time condensate

FlorianDeconinck committed:
Use direct boolean check + final lint

FlorianDeconinck committed:
Merge remote-tracking branch 'NASA/dsl/develop' into dsl/GFDL_1M_single_call

FlorianDeconinck committed:
Restore UW post-merge

FlorianDeconinck committed:
Public API for saturation tables

FlorianDeconinck committed:
Remove stray IDE config file

FlorianDeconinck committed:
Update to new Saturation Tables API

FlorianDeconinck committed:
Restore UW and update to new Saturation Tables API

FlorianDeconinck committed:
Lint

FlorianDeconinck committed:
Update to new Saturation Tables API

FlorianDeconinck committed:
Fix interface for UW with new global wrapper object

FlorianDeconinck committed:
Fix tests suite

FlorianDeconinck committed:
gitignore: vscode

FlorianDeconinck committed:
Orchestration: escape zeros() based on `getattr`

FlorianDeconinck committed:
lint

FlorianDeconinck commented:
Code is validating numerically (C24)

Todo:
- Orchestration is not validating
- Refactor are needed
- Performance is locked behind bad layout on interface translation from Fortran memory

FlorianDeconinck committed:
Update GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/GEOS_MoistGridComp.F90

FlorianDeconinck committed:
Orchestration fixes + setup removed unused temporary 3D
